contracts/secret-storage/SecretManager.sol

```diff
@@ -0,0 +1,731 @@
+    }
+
+    function _removeSelectedEnclave(
+        uint256 _secretId,
+        uint256 _index
+    ) internal {
+        uint256 len = userStorage[_secretId].selectedEnclaves.length;
+        if(_index != len - 1)
+            userStorage[_secretId].selectedEnclaves[_index] = userStorage[_secretId].selectedEnclaves[len - 1];
+        userStorage[_secretId].selectedEnclaves.pop();
+    }
+
+    function _refundExcessDepositAndRemoveStore(
+        uint256 _secretId
+    ) internal {
+        address owner = userStorage[_secretId].owner;
+        uint256 remainingDeposit = userStorage[_secretId].usdcDeposit;
+        delete userStorage[_secretId];
+        USDC_TOKEN.safeTransfer(owner, remainingDeposit);
+    }
+
+    function _updateSecretEndTimestamp(
+        uint256 _secretId,
+        uint256 _endTimestamp,
+        uint256 _usdcDeposit,
+        address _owner
+    ) internal {
+        if(_endTimestamp < block.timestamp)
+            revert SecretManagerInvalidEndTimestamp();
+
+        uint256 currentEndTimestamp = userStorage[_secretId].endTimestamp;
+        if(block.timestamp > currentEndTimestamp)
+            revert SecretManagerAlreadyTerminated();
+
+        if(!_checkIfSecretAcknowledged(_secretId))
+            revert SecretManagerUnacknowledged();
+
+        if(_endTimestamp > currentEndTimestamp) {
+            USDC_TOKEN.safeTransferFrom(_owner, address(this), _usdcDeposit);
+            userStorage[_secretId].usdcDeposit += _usdcDeposit;
+
+            uint256 minUsdcDeposit = (_endTimestamp - userStorage[_secretId].startTimestamp) * userStorage[_secretId].sizeLimit * SECRET_STORE_FEE_RATE * NO_OF_NODES_TO_SELECT;
+            _checkUsdcDeposit(userStorage[_secretId].usdcDeposit, minUsdcDeposit);
+        }
+        else {
+            uint256 removedDuration = currentEndTimestamp - _endTimestamp;
+            uint256 usdcRefund = removedDuration * userStorage[_secretId].sizeLimit * SECRET_STORE_FEE_RATE * NO_OF_NODES_TO_SELECT;
+            
+            userStorage[_secretId].usdcDeposit -= usdcRefund;
+            USDC_TOKEN.safeTransfer(_owner, usdcRefund);
+        }
+
+        userStorage[_secretId].endTimestamp = _endTimestamp;
+
+        emit SecretEndTimestampUpdated(_secretId, _endTimestamp);
+    }
+
+    /**
+     * @dev Checks if the secret has been acknowledged by the first set of selected enclaves.
+     *      Replacement store selected post slashing of the previous store, isn't taken into account.
+     */
+    function _checkIfSecretAcknowledged(
+        uint256 _secretId
+    ) internal view returns (bool) {
+        uint256 len = userStorage[_secretId].selectedEnclaves.length;
+        for (uint256 index = 0; index < len; index++) {
+            if(!_isReplacedStore(_secretId, index) && !userStorage[_secretId].selectedEnclaves[index].hasAcknowledgedStore)
+                return false;
+        }
+        return true;
+    }
+
+    function _terminateSecret(
```

Comment:
need to release the selected secret stores after secret termination, that's missing currently. Also, require it for the case of remove Secret too.

Reply:
yep added